eras/shelley/impl/src/Cardano/Ledger/Shelley/LedgerState.hs

```diff
@@ -1014,10 +1068,149 @@ stakeDistr u ds ps =
+aggregateUtxoCoinByCredential ::
+  forall era.
+  ( Era era
+  ) =>
+  Map Ptr (Credential 'Staking (Crypto era)) ->
+  UTxO era ->
+  Map (Credential 'Staking (Crypto era)) Coin ->
+  Map (Credential 'Staking (Crypto era)) Coin
+aggregateUtxoCoinByCredential ptrs (UTxO u) initial =
+  Map.foldl' accum initial u
+  where
+    accum !ans out =
+      case (getField @"address" out, getField @"value" out) of
+        (Addr _ _ (StakeRefPtr p), c) ->
+          case Map.lookup p ptrs of
+            Just cred -> Map.insertWith (<>) cred (Val.coin c) ans
+            Nothing -> ans
+        (Addr _ _ (StakeRefBase hk), c) ->
+          Map.insertWith (<>) hk (Val.coin c) ans
+        _other -> ans
+
+-- ==============================
+-- operations on IncrementalStake
+
+-- | Incrementally add the inserts 'utxoAdd' and the deletes 'utxoDel' to the IncrementalStake.
+updateStakeDistribution ::
+  ( Era era
+  ) =>
+  IncrementalStake (Crypto era) ->
+  UTxO era ->
+  UTxO era ->
+  IncrementalStake (Crypto era)
+updateStakeDistribution incStake0 utxoDel utxoAdd = incStake2
+  where
+    incStake1 = incrementalAggregateUtxoCoinByCredential id utxoAdd incStake0
+    incStake2 = incrementalAggregateUtxoCoinByCredential invert utxoDel incStake1
+
+-- | Incrementally sum up all the Coin for each staking Credential, use different 'mode' operations
+--   for inserts (id) and deletes (invert). Never store a (Coin 0) balance, since these do not occur
+--   in the non-incremental stye that works directly from the UTxO. This function has a non-incremental
+--   analog 'aggregateUtxoCoinByCredential'
+incrementalAggregateUtxoCoinByCredential ::
+  forall era.
+  ( Era era
+  ) =>
+  (Coin -> Coin) ->
+  UTxO era ->
+  IncrementalStake (Crypto era) ->
+  IncrementalStake (Crypto era)
+incrementalAggregateUtxoCoinByCredential mode (UTxO u) initial =
+  Map.foldl' accum initial u
+  where
+    keepOrDelete new Nothing =
+      case mode new of
+        Coin 0 -> Nothing
+        final -> Just final
+    keepOrDelete new (Just old) =
+      case mode new <> old of
+        Coin 0 -> Nothing
+        final -> Just final
+    accum ans@(IStake stake ptrs) out =
+      let c = Val.coin (getField @"value" out)
+       in case getField @"address" out of
+            Addr _ _ (StakeRefPtr p) -> IStake stake (Map.alter (keepOrDelete c) p ptrs)
+            Addr _ _ (StakeRefBase hk) -> IStake (Map.alter (keepOrDelete c) hk stake) ptrs
+            _other -> ans
+
+-- | Resolve inserts and deletes which were indexed by Ptrs, by looking them up in 'ptrs'
+--   and combining them with the ordinary stake. Zero out the Ptr indexed stake in the result.
+--   This function is meant to only be called at the Epoch boundary in the Snap Rule. It is not meant
+--   to be used to transform the IncrementalStake.
+resolveIncrementalPtrs :: Map Ptr (Credential 'Staking (Crypto era)) -> IncrementalStake (Crypto era) -> IncrementalStake (Crypto era)
```

Comment:
Is there a reason in this case for returning `IncrementalStake` rather than simply `Map Credential Coin`? I assume we zero out the incremental stuff at the epoch boundary anyway?

Reply:
Alternatively, just move this function inside of `incrementalStakeDistr`, which is the only thing to use it anyway.